Index: solr/contrib/dataimporthandler/src/test/org/apache/solr/handler/dataimport/TestContextImpl.java
===================================================================
--- solr/contrib/dataimporthandler/src/test/org/apache/solr/handler/dataimport/TestContextImpl.java	(revision 0)
+++ solr/contrib/dataimporthandler/src/test/org/apache/solr/handler/dataimport/TestContextImpl.java	(revision 0)
@@ -0,0 +1,61 @@
+/**
+ * 
+ */
+package org.apache.solr.handler.dataimport;
+
+import java.util.HashMap;
+
+import org.apache.solr.handler.dataimport.DataConfig.Entity;
+import org.apache.solr.handler.dataimport.DataImporter.RequestParams;
+import org.junit.Test;
+/**
+ * @author fwe
+ *
+ */
+public class TestContextImpl extends AbstractDataImportHandlerTestCase {
+  
+  @Test
+  public void testEntityScope() {
+    ContextImpl ctx = new ContextImpl(new Entity(), new VariableResolverImpl(), null, "something", new HashMap<String,Object>(), null, null);
+    String lala = new String("lala");
+    ctx.setSessionAttribute("huhu", lala, Context.SCOPE_ENTITY);
+    Object got = ctx.getSessionAttribute("huhu", Context.SCOPE_ENTITY);
+    
+    assertEquals(lala, got);
+    
+  }
+  @Test
+  public void testCoreScope() {
+    DataImporter di = new DataImporter();
+    di.loadAndInit("<dataConfig><document /></dataConfig>");
+    DocBuilder db = new DocBuilder(di, new SolrWriter(null, null),new SimplePropertiesWriter(), new RequestParams());
+    ContextImpl ctx = new ContextImpl(new Entity(), new VariableResolverImpl(), null, "something", new HashMap<String,Object>(), null, db);
+    String lala = new String("lala");
+    ctx.setSessionAttribute("huhu", lala, Context.SCOPE_SOLR_CORE);
+    Object got = ctx.getSessionAttribute("huhu", Context.SCOPE_SOLR_CORE);
+    assertEquals(lala, got);
+    
+  }
+  @Test
+  public void testDocumentScope() {
+    ContextImpl ctx = new ContextImpl(new Entity(), new VariableResolverImpl(), null, "something", new HashMap<String,Object>(), null, null);
+    ctx.setDoc(new DocBuilder.DocWrapper());
+    String lala = new String("lala");
+    ctx.setSessionAttribute("huhu", lala, Context.SCOPE_DOC);
+    Object got = ctx.getSessionAttribute("huhu", Context.SCOPE_DOC);
+    
+    assertEquals(lala, got);
+    
+  }
+  @Test
+  public void testGlobalScope() {
+    ContextImpl ctx = new ContextImpl(new Entity(), new VariableResolverImpl(), null, "something", new HashMap<String,Object>(), null, null);
+    String lala = new String("lala");
+    ctx.setSessionAttribute("huhu", lala, Context.SCOPE_GLOBAL);
+    Object got = ctx.getSessionAttribute("huhu", Context.SCOPE_GLOBAL);
+    
+    assertEquals(lala, got);
+    
+  }
+  
+}
Index: solr/contrib/dataimporthandler/src/java/org/apache/solr/handler/dataimport/ContextImpl.java
===================================================================
--- solr/contrib/dataimporthandler/src/java/org/apache/solr/handler/dataimport/ContextImpl.java	(revision 1201784)
+++ solr/contrib/dataimporthandler/src/java/org/apache/solr/handler/dataimport/ContextImpl.java	(working copy)
@@ -155,7 +155,7 @@
 
   private void putVal(String name, Object val, Map map) {
     if(val == null) map.remove(name);
-    else entitySession.put(name, val);
+    else map.put(name, val);
   }
 
   @Override
